crates/rust-analyzer/src/config.rs

```diff
@@ -651,23 +654,304 @@ pub struct Config {
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            ConfigNodeKey::Ratoml(source_root_id) => serializer.serialize_u32(source_root_id.0),
+            ConfigNodeKey::Client => serializer.serialize_str("client"),
+            ConfigNodeKey::User => serializer.serialize_str("user"),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+enum ConfigNodeValue<'a> {
+    /// `rust-analyzer::config` module works by setting
+    /// a mapping between `SourceRootId` and `ConfigInput`.
+    /// Storing a `FileId` is mostly for debugging purposes.
+    Ratoml(&'a RatomlNode),
+    Client(&'a FullConfigInput),
+}
+
+impl Config {
+    /// FIXME @alibektas : Before integration tests, I thought I would
+    /// get the debug output of the config tree and do assertions based on it.
+    /// The reason why I didn't delete this is that we may want to have a lsp_ext
+    /// like "DebugConfigTree" so that it is easier for users to get a snapshot of
+    /// the config state for us to debug.
     #[allow(dead_code)]
-    parent: Option<SourceRootId>,
+    /// Walk towards the root starting from a specified `ConfigNode`
+    fn traverse(
+        &self,
+        start: ConfigNodeKey,
+    ) -> impl Iterator<Item = (ConfigNodeKey, ConfigNodeValue<'_>)> {
+        let mut v = vec![];
+
+        if let ConfigNodeKey::Ratoml(start) = start {
+            let mut par: Option<SourceRootId> = Some(start);
+            while let Some(source_root_id) = par {
+                par = self.source_root_parent_map.get(&start).copied();
+                if let Some(config) = self.ratoml_files.get(&source_root_id) {
+                    v.push((
+                        ConfigNodeKey::Ratoml(source_root_id),
+                        ConfigNodeValue::Ratoml(config),
+                    ));
+                }
+            }
+        }
+
+        v.push((ConfigNodeKey::Client, ConfigNodeValue::Client(&self.client_config.node)));
+
+        if let Some(user_config) = self.user_config.as_ref() {
+            v.push((ConfigNodeKey::User, ConfigNodeValue::Ratoml(user_config)));
+        }
+
+        v.into_iter()
+    }
+
+    pub fn user_config_path(&self) -> &VfsPath {
+        &self.user_config_path
+    }
+
+    pub fn should_update(&self) -> bool {
+        self.should_update
+    }
+
+    // FIXME @alibektas : Server's health uses error sink but in other places it is not used atm.
+    pub fn apply_change(&self, change: ConfigChange, error_sink: &mut ConfigError) -> Config {
+        let mut config = self.clone();
+        let mut toml_errors = vec![];
+        let mut json_errors = vec![];
+
+        config.should_update = false;
+
+        if let Some((file_id, change)) = change.user_config_change {
+            config.user_config = Some(RatomlNode {
+                file_id,
+                node: GlobalLocalConfigInput::from_toml(
+                    toml::from_str(change.to_string().as_str()).unwrap(),
```

Comment:
Nit

```suggestion
                    toml::from_str(&change.to_string()).unwrap(),
```

more importantly though, that unwrap is a bad idea, I am fairly sure this code path can be reached with an invalid toml file